tests/test_tuning_XXXMBE.cpp

```diff
@@ -0,0 +1,157 @@
+
+#define BOOST_TEST_MODULE TestTuningXXXMBE
+#include <boost/test/unit_test.hpp>
+
+struct Column : public std::vector<std::string> {
+		Column(const std::string& name_, const int size = 0, const int num_rows_estimate = 1000) : 
+				std::vector<std::string>(size), name(name_)
+		{
+				this->reserve(num_rows_estimate);
+		}
+
+		// Return vector of double values, invalid elements set to NaN
+		std::vector<double> dvalues() const {
+				std::vector<double> vec;
+				vec.reserve(this->size());
+				
+				const auto& conv_func = [](const std::string& strval) {
+						double dval;
+						try {
+								dval = std::stod(strval);
+						} catch (std::invalid_argument& exc) {
+								dval = std::numeric_limits<double>::quiet_NaN();
+						}
+						return dval;
+				};
+
+				std::transform(this->cbegin(), this->cend(), std::back_inserter(vec), conv_func);
+				return vec;
+		}
+
+		// Return vector of double values values, invalid elements set to std::numeric_limits<int>::min()
+		std::vector<int> ivalues() const {
+				std::vector<int> vec;
+				vec.reserve(this->size());
+				
+				const auto& conv_func = [](const std::string& strval) {
+						int ival;
+						try {
+								ival = static_cast<int>(std::stod(strval));
+						} catch (std::invalid_argument& exc) {
+								ival = std::numeric_limits<int>::min();
+						}
+						return ival;
+				};
+
+				std::transform(this->cbegin(), this->cend(), std::back_inserter(vec), conv_func);
+				return vec;
+		}
+
+		std::string name;
+};
+
+
+struct ColumnData {
+		ColumnData(const std::string& file_name, const int num_columns_estimate=20) {
+				raw_columns.reserve(num_columns_estimate);
+				load_file(file_name);
+		}
+
+		void load_file(const std::string& file_name) {
+				// Open file and read first line with column names
+				std::ifstream ifs(file_name);
+				std::string line, colname;
+
+				std::getline(ifs, line);
+				std::istringstream iss(line);
+				while (iss >> colname) {
+						column_names.push_back(colname);
+						raw_columns.emplace_back(colname);
+						columns[colname] = &(raw_columns.back());
+				}
+				const int num_columns = column_names.size();
+
+				// Read remaining lines into std::string vectors
+				int lineno = 1;
+				while (std::getline(ifs, line)) {
+						iss.str(line); iss.clear();
+						int i=0;
+						while (iss >> colname && i < num_columns) {
+								raw_columns[i].push_back(colname);
+								++i;
+						}
+						if (i > num_columns) {
+								std::cout << "Warning: Ignoring extra column(s) on line " << lineno << std::endl;
+						}
+						++lineno;
+				}
+		}
+
+		// Get data vectors of different types
+		std::vector<double> get_dvector(const std::string& colname) { return columns[colname]->dvalues(); }
+		std::vector<int> get_ivector(const std::string& colname) { return columns[colname]->ivalues(); }
+		// Default is to return double values
+		std::vector<double> operator[](const std::string& colname) { return columns[colname]->dvalues(); }
```

Comment:
Strictly speaking, `columns[colname]` will always (attempt to) insert a new node into the `columns` map.  Maybe that's fine for your purposes here, but if it's not then we should probably write these as
```C++
columns.at(colname)->...
```
instead.  Then you could also declare the member functions as `const`.

Reply:
Good point, switching to at.